### DISCUSSION TITLE
Specifying a shader pipeline within Kage code

### DISCUSSION THREAD

**Discussion by bandaloo** [hajimehoshi/ebiten#1798] [Ideas]
Hi, I'm very interested in how ebiten deals with fragment shaders, and how you implemented a go-like DSL for cross-compiling to different shading languages. I happened to create a similar thing, called tinsl, which is a sort of extension to GLSL with type inference using `:=`. Beyond this,  I was also interested in being able to render to texture half-way through a shader, and read from that texture later in the shader, essentially specifying an entire rendering pipeline in shader-like code. The tinsl compiler will split this up into multiple shaders. Consider this example of doing a "bloom" effect:

```glsl
def threshold 0.5 // threshold for luma to mask off

fn luma(vec4 color) { return dot(color.rgb, vec3(0.299, 0.587, 0.114)); }

// linear gaussian blur; we do a horizontal then vertical blur in two separate shader
// passes to get a radial blur, which is much more efficient than a single pass
fn blur(vec2 direction, int channel) {
  uv := npos;
  off1 := vec2(1.3333333333333333) * direction;
  mut color := vec4(0.);
  color += frag(uv, channel) * 0.29411764705882354;
  color += frag(uv + (off1 / res), channel) * 0.35294117647058826;
  color += frag(uv - (off1 / res), channel) * 0.35294117647058826;
  return color;
}

// reusable "procedure" for doing a horizontal blur, then vertical, then repeating it
pr two_pass_blur(float size, int reps, int channel = -1) {
  loop reps {
    blur(vec2(size, 0.), channel); refresh;
    blur(vec2(0., size), channel);
  }
}

// mask off the bright fragments, then render out to off-screen texture 1
{ vec4(frag.rgb * (step(1. - luma(frag0), 1. - threshold)), frag.a); } -> 1

// read from texture 1, blur it three times then render back to texture 1
1 -> { @two_pass_blur(size: 1., reps: 3);} -> 1

// add texture 0 (the original image) and texture 1 (the masked blurred image)
// and implicitly render to screen
{ frag0 + frag1; }
```

This gets compiled into four separate GLSL fragment shaders (a sort of "shader tree") which then get run in a specific sequence. I think such a feature would work really well within ebiten, since you already have a shader DSL. This would enable the user to do much more complex post-processing effects that cannot be accomplished efficiently within a single shader. (Like Kage, tinsl also doesn't have structs yet, because parsing is a lot of work lol)

**Comment by Zyko0** [comment]
For the moment I think you can handle the N shader invocations in the order you want, in a user function responsible for the post-processing effects, right ?
I'm not sure, I understood the benefits of your suggestion, is it for better shader management ? Like they would be bound already, instead of having to give them specific folder and a specific go-function to handle the multiple calls in order ?

**Reply by bandaloo** [reply]
You wouldn't have to give them a specific Go function; the information for handling the order of the shaders is baked into the semantics of Tinsl itself. Otherwise you're totally right, although I think it would just be cool and unique among game libraries. One advantage is that it does get rid of a lot of repeated code. For example, shaders can't share code with one another and the way this is usually dealt with in game engines is with some kind of `#include` preprocessor macro hacked into the language (there's no such thing in GLSL proper), or some metaprogramming with string concatenation done in the host language. The compiled output (I edited the whitespace) of the above code gets generated into these four GLSL files:

```glsl
// new file: masking shader 
#version 300 es
precision mediump float;
out vec4 fragColor;
uniform sampler2D uSampler0;
uniform vec2 uResolution;
#define THRESHOLD .5

float luma(vec4 color) {
  return dot(color.rgb, vec3(.299,.587,.114));
}

void main() {
  vec2 uv = gl_FragCoord.xy / uResolution;
  vec3 col = step(1. - luma(texture(uSampler0, uv)), 1. - THRESHOLD);
  fragColor= vec4(texture(uSampler0, uv).rgb * col, 1.);
}
```

```glsl
// new file: horizontal blur shader
#version 300 es
precision mediump float;
out vec4 fragColor;
uniform sampler2D uSampler1;
uniform vec2 uResolution;

vec4 blur(vec2 direction, sampler2D channel) {
  vec2 uv = gl_FragCoord.xy / uResolution;
  vec2 off1 = vec2(1.3333333333333333) * direction;
  vec4 color =vec4(0.);
  color += texture(channel, uv)*0.29411764705882354;
  color += texture(channel, uv + off1/uResolution) * 0.35294117647058826;
  color += texture(channel, uv - off1/uResolution) * 0.35294117647058826;
  return color;
}

void main() {
  fragColor = blur(vec2(1., 0.), uSampler1);
}
```

```glsl
// new file: vertical blur shader
#version 300 es
precision mediump float;
out vec4 fragColor;
uniform sampler2D uSampler1;
uniform vec2 uResolution;

vec4 blur(vec2 direction, sampler2D channel){
  vec2 uv = gl_FragCoord.xy / uResolution;
  vec2 off1 = vec2(1.3333333333333333) * direction;
  vec4 color =vec4(0.);
  color += texture(channel, uv) * 0.29411764705882354;
  color += texture(channel, uv + off1/uResolution) * 0.35294117647058826;
  color += texture(channel, uv - off1/uResolution) * 0.35294117647058826;
  return color;
}

void main() {
  fragColor = blur(vec2(0.,1.), uSampler1);
}
```

```glsl
// new file: copy texture 1 onto texture 2
#version 300 es
precision mediump float;
out vec4 fragColor;
uniform sampler2D uSampler0;
uniform sampler2D uSampler1;
uniform vec2 uResolution;

void main() {
  vec2 uv = gl_FragCoord.xy / uResolution;
  fragColor = texture(uSampler0, uv) + texture(uSampler1, uv);
}
```

Sorry for the wall of code. Beyond this, the correct amount of off-screen textures for RTT effects will be created (probably with `NewImage` in ebiten.) These strings of fragment shader source code get returned in a tree structure so sections can be looped over. The reason I think this would make sense for ebiten is that ebiten already has a custom DSL for shader portability; it could be extended to include meta-information about how to handle a rendering pipeline of shaders within the shader-like code itself. And, you could code generate the shader pipeline meta-language to completely re-architect the rendering pipeline on the fly, which you wouldn't be able to do easily if the rendering pipeline logic were handled by Go passing around off-screen textures and swapping shaders in a fixed way.

**Reply by hajimehoshi** [reply]
Thank you for posting!

So, your idea is basically enabling a complex pipeline involving multiple offscreen textures in one shader program, right? This seems very interesting, but probably Ebiten won't adopt this as this would not match with the current API model and implementation so far.

I appreciate you posted your idea in this discussion anyway. Opinions or suggestions from various perspectives are very insightful.

**Reply by bandaloo** [reply]
Yes, that's exactly right! I figured such a thing might not be a perfect fit for Ebiten but it's fun to think about. Thanks for taking the time to read my thoughts about it.

**Reply by hajimehoshi** [reply]
BTW,

```
1 -> { @two_pass_blur(size: 1., reps: 3);} -> 1
```

This seems the rendering source and target are the same. Does this really work, or does `@two_pass_blur` uses an offscreen inside?

**Reply by bandaloo** [reply]
Good question! Yeah, it's not really the same texture since you can't read from and write to the same texture, like you pointed out. In my implementation it renders to a separate texture (call it "tex back") and then tex back becomes the new texture 1. Textures get swapped around like this and it works because they are all the same size. But the user doesn't have to know that, and they can pretend like they are reading to and writing from the same texture. I made an online playground. (Beware there are probably some bugs) https://bandaloo.fun/playground

**Reply by hajimehoshi** [reply]
OK, thank you for elaborating!
